pyrad/datatypes/leaf.py

```diff
@@ -0,0 +1,367 @@
+                    terms['src'] = 16 * b'\x00'
+                if terms['dst'] == b'\x00\x00\x00\x00':
+                    terms['dst'] = 16 * b'\x00'
+            elif key == 'action' and value == 'accept':
+                terms[key] = b'\x01'
+            elif key == 'action' and value == 'redirect':
+                terms[key] = b'\x20'
+            elif key == 'direction' and value == 'out':
+                terms[key] = b'\x00'
+            elif key in ('src', 'dst'):
+                if family == 'ipv4':
+                    ip = IPv4Network(value)
+                else:
+                    ip = IPv6Network(value)
+                terms[key] = ip.network_address.packed
+                terms[key + 'l'] = struct.pack('B', ip.prefixlen)
+            elif key in ('sport', 'dport'):
+                terms[key] = struct.pack('!H', int(value))
+            elif key in ('sportq', 'dportq', 'proto'):
+                terms[key] = struct.pack('B', int(value))
+
+        trailer = 8 * b'\x00'
+
+        result = b''.join(
+            (terms['family'], terms['action'], terms['direction'], b'\x00',
+             terms['src'], terms['dst'], terms['srcl'], terms['dstl'],
+             terms['proto'], b'\x00',
+             terms['sport'], terms['dport'], terms['sportq'], terms['dportq'],
+             b'\x00\x00', trailer))
+        return result
+
+    def decode(self, raw):
+        return raw
+
+    def print(self, attribute, decoded):
+        return decoded
+
+class Byte(AbstractLeaf):
+    """
+    leaf datatype class for bytes
+    """
+    def __init__(self):
+        super().__init__('byte')
+
+    def encode(self, attribute, decoded):
+        try:
+            num = int(decoded)
+        except Exception as exc:
+            raise TypeError('Can not encode non-integer as byte') from exc
+        return struct.pack('!B', num)
+
+    def decode(self, raw):
+        return struct.unpack('!B', raw)[0]
+
+    def print(self, attribute, decoded):
+        return decoded
+
+class Date(AbstractLeaf):
+    """
+    leaf datatype class for dates
+    """
+    def __init__(self):
+        super().__init__('date')
+
+    def encode(self, attribute, decoded):
+        if not isinstance(decoded, int):
+            raise TypeError('Can not encode non-integer as date')
+        return struct.pack('!I', decoded)
+
+    def decode(self, raw):
+        return (struct.unpack('!I', raw))[0]
+
+    def print(self, attribute, decoded):
+        return decoded
+
+class Ether(AbstractLeaf, ABC):
+    """
+    leaf datatype class for ethernet addresses
+    """
+    def __init__(self):
+        super().__init__('ether')
+
+    def encode(self, attribute, decoded):
+        return struct.pack('!6B', *map(lambda x: int(x, 16), decoded.split(':')))
+
+    def decode(self, raw):
+        ':'.join(map('{0:02x}'.format, struct.unpack('!6B', raw)))
+
+    def print(self, attribute, decoded):
+        return decoded
```

Comment:
I don't understand what the "print" functions are supposed to do here.

The requirement for `print` is "print this value to an ASCII string".  The class also needs a related `parse` routine.  which takes an ASCII string, and returns a python structure / dict containing the attribute

This functionality lets us do the following tests:

decode HEX -> python stuff
print python stuff - ASCII

Human reads the ASCII and manually checks it agains the HEX to be sure it's OK.

We can then do

parse ASCII -> python stuff
encode python stuff -> HEX

which should then result in the same HEX output as above.

Once the unit tests have been verified manually (or copied from FreeRADIUS), the tests can be run in CI.  The tests then ensure that:

a) the decoder can handle the packets correctly
b) the ASCII output doesn't change
c) the parser can read ASCII output
d) the encoder can encode things correctly

Reply:
So yes, the `print` function is returning an ASCII string. It just so happens that Pyrad has been storing these datatypes as strings, thus I simply return the value (which is an ASCII string).